### PR TITLE
Rdm 7621 - Migrate Data Store to OIDC

### PR DESCRIPTION
**Pull request by smathangi**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-7621

### Change description ###

Migrate Data Store to OIDC

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java [hunk 1]
```diff
@@ -1,63 +1,102 @@
...
             .authorizeRequests()
             .anyRequest()
-            .authenticated();
+            .authenticated()
+            .and()
+            .oauth2ResourceServer()
+            .jwt()
+            .jwtAuthenticationConverter(jwtAuthorityExtractor)
+            .and()
+            .and()
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java]: duplicate?

File: src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java [hunk 2]
```diff
@@ -1,63 +1,102 @@
...
 
-    private final AuthCheckerServiceAndUserFilter authCheckerFilter;
+    @Value("${spring.security.oauth2.client.provider.oidc.issuer-uri}")
+    private String issuerUri;
+
+    @Value("${oidc.issuer}")
+    private String issuerOverride;
+
+    private final ServiceAuthFilter serviceAuthFilter;
+    private final PathParamSecurityFilter pathParamSecurityFilter;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java]: this filter is only used for v1 endpoints correct? if yes I'd suggest we make that more explicit in the name, like
`V1EndpointsSecurityFilter` or `V1EndpointsPathParamSecurityFilter`

File: src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java [hunk 3]
```diff
@@ -19,10 +22,10 @@
     @Primary
     @Bean(name = "DefaultObjectMapper")
     public ObjectMapper defaultObjectMapper() {
-        ObjectMapper mapper = new ObjectMapper();
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java]: this is the default mapper. Shall we ask some regression testing around this change?

File: src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java [hunk 4]
```diff
@@ -19,10 +22,10 @@
...
-        mapper.registerModules(new JavaTimeModule());
-        mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
-        return mapper;
+        return new ObjectMapper()
+            .registerModule(new Jdk8Module())
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java]: this is the default mapper. Shall we ask some regression testing around this change?

File: src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java [hunk 5]
```diff
@@ -26,35 +31,42 @@ public HttpHeaders authorizationHeaders() {
...
     public HttpHeaders userAuthorizationHeaders() {
-        final ServiceAndUserDetails serviceAndUser = (ServiceAndUserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
         final HttpHeaders headers = new HttpHeaders();
-        headers.add(HttpHeaders.AUTHORIZATION, serviceAndUser.getPassword());
+        headers.add(HttpHeaders.AUTHORIZATION, getUserBearerToken());
         return headers;
     }
 
+    public UserInfo getUserInfo() {
+        return idamRepository.getUserInfo(getUserToken());
```
Comment (nikola-naydenov-hmcts) [inline src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java]: Does this work? 
image
You aren't getting the bearer token, just the token and the endpoint expect bearer. I checked the client and it doesnt seem to prepend "Bearer" to the header

File: src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    public static final String AUTHORISATION = "ServiceAuthorization";
+
+    private final List<String> authorisedServices;
+    private final AuthTokenValidator authTokenValidator;
+
+    public ServiceAuthFilter(AuthTokenValidator authTokenValidator, List<String> authorisedServices) {
+        this.authTokenValidator = authTokenValidator;
+        if (authorisedServices.isEmpty()) {
+            throw new IllegalArgumentException("Must have at least one service defined");
+        }
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java]: do we need to throw an exception here? if the list of authorised services is empty the request simply won't be allowed

File: src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.function.Function;
+
+@Component
```
Comment (nikola-naydenov-hmcts) [inline src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java]: Where is this bean instantiated?
I cant seem to find that anywhere in this code

File: src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java (hunk 1 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/SecurityConfiguration.java]: nope it's required.

File: src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java (hunk 4 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/config/JacksonObjectMapperConfig.java]: tbh not much change there - just added to support ParameterNamesModule. Yes, all this work needs regression testing.

File: src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java (hunk 5 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java]: "Bearer" append in the  idamRepository  method https://github.com/hmcts/ccd-data-store-api/blob/9fdc0723f8c35c21b6d73e9da9d52fc5c28cdd7b/src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java#L21

File: src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java (hunk 7 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java]: https://github.com/hmcts/ccd-data-store-api/blob/9fdc0723f8c35c21b6d73e9da9d52fc5c28cdd7b/src/main/java/uk/gov/hmcts/ccd/AuthCheckerConfiguration.java#L51-L52

File: src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    private String extractBearerToken(HttpServletRequest request) {
+        String token = request.getHeader(AUTHORISATION);
+        if (token == null) {
+            throw new UnauthorizedException("ServiceAuthorization Token is missing");
+        }
+        return token.startsWith("Bearer ") ? token : "Bearer " + token;
+    }
+
+
+}
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java]: I'm a bit concerned about CCD having to implement from scratch S2S authorisation. It feels like reimplementing the wheel. I was hoping the Programme has a lib we can use for this. 
@timja any suggestion?

File: src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java (hunk 6 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java]: Filter bean creation itself fails if authorised list is not configured (empty)

File: src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        throws ServletException, IOException {
+
+        Collection<String> authorizedRoles = authorizedRolesExtractor.apply(request);
+        Optional<String> userIdOptional = userIdExtractor.apply(request);
+
+        if (!authorizedRoles.isEmpty() || userIdOptional.isPresent()) {
+            try {
+                verifyRoleAndUserId(authorizedRoles, userIdOptional);
+            } catch (UnauthorizedException ex) {
+                LOG.warn("Unauthorised roles or userId in the request path", ex);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/security/filters/PathParamSecurityFilter.java]: can we have separate error messages? would make issue investigation simpler

File: src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+public class IdamRepository {
+
+    private final IdamClient idamClient;
+
+    @Autowired
+    public IdamRepository(IdamClient idamClient) {
+        this.idamClient = idamClient;
+    }
+
+    @Cacheable(value = "userInfoCache")
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java]: do we know for how long we are caching?

Comment (smathangi) [reply src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java]: same as current setting 1440 seconds - 4 hours. A lot tbh. Also, check my comment on DefaultUserRepository

File: src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java (hunk 5 above)
Comment (nikola-naydenov-hmcts) [reply src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java]: ah I see

File: src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java (hunk 10 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/security/idam/IdamRepository.java]: mmm yes a bit too much

File: src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
+package uk.gov.hmcts.ccd.security.filters;
+
+import org.slf4j.Logger;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/security/filters/ServiceAuthFilter.java]: glad to see that this has now been included in https://github.com/hmcts/service-auth-provider-java-client/pull/46.
I guess now we are going to remove this and take it from `service-auth-provider-java-client` right?

File: src/main/java/uk/gov/hmcts/ccd/security/JwtAuthorityExtractor.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    private final IdamRepository idamRepository;
+
+    @Autowired
+    public JwtAuthorityExtractor(IdamRepository idamRepository) {
+        this.idamRepository = idamRepository;
+    }
+
+    @Override
+    protected Collection<GrantedAuthority> extractAuthorities(Jwt jwt) {
```
Comment (mrganeshraja) [inline src/main/java/uk/gov/hmcts/ccd/security/JwtAuthorityExtractor.java]: @smathangi as discussed, lets not override this deprecated method and instead change to `convert`

File: charts/ccd-data-store-api/Chart.yaml [hunk 13]
```diff
@@ -2,7 +2,7 @@ description: Helm chart for the HMCTS CCD Data Store
 name: ccd-data-store-api
 apiVersion: v1
 home: https://github.com/hmcts/ccd-data-store-api
-version: 1.5.1
+version: 1.5.2
```
Comment (nikola-naydenov-hmcts) [inline charts/ccd-data-store-api/Chart.yaml]: Considering the gravity of the changes, maybe 1.6?

Comment (mario-paniccia) [reply charts/ccd-data-store-api/Chart.yaml]: 👍

**Pull request closed without merging**